internal/threadgroup/threadgroup.go

```diff
@@ -0,0 +1,96 @@
+package threadgroup
+
+import (
+	"context"
+	"errors"
+	"sync"
+)
+
+type (
+	// A ThreadGroup provides synchronization between a module and its
+	// goroutines to enable clean shutdowns
+	ThreadGroup struct {
+		mu     sync.Mutex
```

Comment:
I don't think so since we need to close the channel and wait on the waitgroup atomically. Removing it would introduce a race between `Add()` and `Stop()`